### PR TITLE
Filter & Order ovehaul

### PR DESCRIPTION
**Pull request by Kitefiko**
### Filters
 - reworked custom filter methods to have enforce params and use `Q` object
 - added DISTINCT
 - more granular `FilterLookups`

### Order
 - custom object or field method with enforced params
 - Ordering enum extended with Nulls related members

## Types of Changes

- [x] Core
- [x] Bugfix
- [x] New feature
- [x] Enhancement/optimization
- [x] Documentation

## Issues Fixed or Closed by This PR

* https://github.com/strawberry-graphql/strawberry-django/issues/399

* https://github.com/strawberry-graphql/strawberry-django/issues/125
* https://github.com/strawberry-graphql/strawberry-django/issues/142
* https://github.com/strawberry-graphql/strawberry-django/issues/236
* https://github.com/strawberry-graphql/strawberry-django/issues/422
* https://github.com/strawberry-graphql/strawberry-django/issues/427
* https://github.com/strawberry-graphql/strawberry-django/issues/459

## Checklist

- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have read the CONTRIBUTING document.
- [x] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

File: docs/guide/ordering.md [hunk 1]
```diff
@@ -33,6 +39,181 @@ input FruitOrder {
...
+    name: auto
+
+    @strawberry_django.order_field
+    def discovered_by(self, value: bool, prefix: str) -> list[str]:
+        if not value:
+            return []
+        return [f"{prefix}discover_by__name", f"{prefix}name"]
+
+    @strawberry_django.order_field
+    def orderer_number(
```
Comment (OdysseyJ) [inline docs/guide/ordering.md:58]: orderer_number -> order_number
maybe wrong typo?

Comment (Kitefiko) [reply docs/guide/ordering.md]: You are completely correct, good catch. Fixed.

**Comment by bellini666** [comment]
hey @Kitefiko ,

Let me know once this is finished and you want me to review it :)

**Comment by Kitefiko** [comment]
Hey, 
I think it is pretty much done, so review away :)

There is one decision to be made that I mentioned in original issue:

 - We could ignore `null` even for `exact`, `iexact`, make user use `isnull` (I would argue, this is better anyway). This was mentioned here -  https://github.com/strawberry-graphql/strawberry-django/issues/459

**Review by bellini666 — CHANGES_REQUESTED** [review]
@Kitefiko thank you _very much_ for this! :)

And sorry for taking too long for the first review. It is a very large PR and I was lacking time to sit and review it.

I think you did a great job in general! There are a couple of comments to address, mostly nitpicks, and I think it should be good to merge.

Also make sure to rebase/merge on top of `main` and check any typing issues.

From now on, next reviews should be faster

**Comment by bellini666** [comment]
> Hey, I think it is pretty much done, so review away :)
> 
> There is one decision to be made that I mentioned in original issue:
> 
> * We could ignore `null` even for `exact`, `iexact`, make user use `isnull` (I would argue, this is better anyway). This was mentioned here -  [lookups doesn't accept null #459](https://github.com/strawberry-graphql/strawberry-django/issues/459)

Yes, I agree that we should be used `isnull` here

**Comment by Kitefiko** [comment]
Hi,

resolved all raised issues. Additionally  made all lookups ignore `null` by default.

Would be nice to have some pylance pre-commit hook for typing, if there is one?

File: docs/guide/filters.md [hunk 2]
```diff
@@ -17,6 +17,12 @@ class Fruit:
     ...
 ```
 
+!!! tip
+
+    In most cases filter fields should have `Optional` annotations and default value `strawberry.UNSET`.
```
[the remark this answers is not in the document]
Comment (Kitefiko) [reply docs/guide/filters.md]: Agreed. Adjusted.

File: tests/filters/test_filters_v2.py [hunk 3]
```diff
@@ -0,0 +1,368 @@
...
+    class Filter:
+        @strawberry_django.order_field
+        def custom_filter(self, root, info, prefix, value: auto, queryset):
+            assert self == _filter, "Unexpected self passed"
+            assert root == _filter, "Unexpected root passed"
+            assert info == _info, "Unexpected info passed"
+            assert prefix == "ROOT", "Unexpected prefix passed"
+            assert value == "SOMETHING", "Unexpected value passed"
+            assert queryset == _queryset, "Unexpected queryset passed"
+            raise Exception("WAS CALLED")
```
[the remark this answers is not in the document]
Comment (Kitefiko) [reply tests/filters/test_filters_v2.py]: Here I adjusted test to be on returned value of q_object or order_list. It is a more readable for sure.

File: strawberry_django/fields/filter_types.py [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+_SKIP_MSG = ". Filter will be skipped on `null` value"
+
+
+@strawberry.input
+class BaseFilterLookup(Generic[T]):
+    exact: Optional[T] = filter_field(
+        filter_none=True,
+        description="Exact match. If `null` value is used filter is equivalent to `isNull=True`",
+    )
+    is_null: Optional[bool] = filter_field(description=f"Assignment test{_SKIP_MSG}")
```
[the remark this answers is not in the document]
Comment (Kitefiko) [reply strawberry_django/fields/filter_types.py]: adjusted

File: tests/filters/test_filters_v2.py [hunk 5]
```diff
@@ -0,0 +1,368 @@
...
+            [1, "str", GlobalID("", "24"), Version.THREE],
+            [1, "str", "24", Version.THREE.value],
+        ),
+    ],
+)
+def test_resolve_value(value, resolved):
+    assert _resolve_value(value) == resolved
+
+
+def test_filter_field_validation():
```
[the remark this answers is not in the document]
Comment (Kitefiko) [reply tests/filters/test_filters_v2.py]: separated

File: strawberry_django/fields/filter_types.py [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+    in_list: Optional[List[T]] = filter_field(
+        description=f"Exact match of items in a given list{_SKIP_MSG}"
+    )
+
+
+@strawberry.input
+class RangeLookup(Generic[T]):
+    left: Optional[T] = None
+    right: Optional[T] = None
```
[the remark this answers is not in the document]
Comment (Kitefiko) [reply strawberry_django/fields/filter_types.py]: yep, seems better. adjusted

**Comment by bellini666** [comment]
> Hi,
> 
> resolved all raised issues. Additionally made all lookups ignore `null` by default.

Awesome! Will take a look at the end of the day for me here :)

> Would be nice to have some pylance pre-commit hook for typing, if there is one?

Hrm, maybe we can do something like this: https://github.com/microsoft/pyright/blob/main/docs/ci-integration.md#running-pyright-as-a-pre-commit-hook

The reason why I didn't add that yet is because `pyright` takes a lot more time to run than other general linters

**Comment by Kitefiko** [comment]
> Hrm, maybe we can do something like this: https://github.com/microsoft/pyright/blob/main/docs/ci-integration.md#running-pyright-as-a-pre-commit-hook
> 
> The reason why I didn't add that yet is because `pyright` takes a lot more time to run than other general linters

Makes sense, maybe adding it to pre-commit (for developers), skip it during github pre-commit & then run it explicitly same as now?

**Review by bellini666 — APPROVED** [review]
Thank you very much for all the work you put into this! :)

Going to merge and release this with a `BREAKING CHANGE` note in the release notes

**Pull request merged**